File: fuse/examples/python/actcancel/fccli.py
import os
import sys
import json
import io
import select
import errno

import time
from iolib import *

def fctest() :
    error = 0
    svcdir = sys.argv[1]

    idx = 123456
    try:
        stmFile = svcdir + "/streams/stream_0"
        stmfd = os.open( stmFile, os.O_WRONLY | os.O_CREAT | os.O_DIRECT )
        stmfp = os.fdopen(stmfd, "wb")

        for j in range( 101 ) :
            greeting = "hello, world" + str( j ) + "\n"

            try:
                os.write( stmfd, greeting.encode("UTF-8"))
            except Exception as err:
                #sending too fast
                #wait for peer to consume them by selecting the fd
                if err.errno == errno.EAGAIN :
                    print( "flow control @loop %d" % j )
                    outNotify = [stmfp]
                    notifyList = select.select([], outNotify, [])
                    j-=1
                    continue
                break

        stmfd.close()
    except Exception as err:
        print( err )
        print( "usage: fctest.py <service path>")
        print( "\t<service path> is /'path to mountpoint'/connection_X/ActiveCancel")
        stmfd.close()
    finally :
        return error


def main() :
    fctest()

if __name__ == "__main__":
    main()

File: fuse/examples/cpp/actcancel/fccli.cpp
//compile: g++ -o fccli -ggdb -std=c++11 -lstdc++ -ldl fccli.cpp
#include "stdio.h"
#include <string>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>

int main( int argc, char** argv )
{
    if( argc < 2 )
        return EINVAL;

    int ret = 0;
    std::string strPath = argv[1];
    strPath += "/streams/stream_0";
    int fd = open( strPath.c_str(), O_CREAT | O_WRONLY, 0600 );
    if( fd < 0 )
    {
        perror(strerror(fd));
        printf( "usage: fctest.py <service path>\n");
        printf( "\t<service path> is /'path to mountpoint'/connection_X/ActiveCancel\n");
        return fd;
    }
    for( int i = 0; i < 10; i++)
        for(int j = 0; j < 101; j++ )
        {
            std::string strMsg = "Hello, World ";
            strMsg += std::to_string(j+1000*i);
            strMsg.push_back('\n');
            ret = write( fd, strMsg.c_str(), strMsg.size() );
            if( ret >= 0 )
                continue;
            ret = errno;
            if( ret == EAGAIN )
            {
                fd_set wfds;
                FD_ZERO( &wfds );
                FD_SET( fd, &wfds );
                select(fd+1, nullptr, &wfds, nullptr, nullptr );
                j--;
            }
            break;
        }

    close(fd);
    return ret;
}